File: Ruby/009_string_each_char_occurrence.rb
######################################################################
# this file contains different implementations of counting how many
# times each unique character appears in string
######################################################################

def chars_counter_1(string)
  unique_chars_hash = {}
  for char in string.chars.to_a.uniq
    unique_chars_hash[char] = string.count(char)
    string.delete(char)
  end
  unique_chars_hash
end

def chars_counter_2(string)
  unique_chars_hash = {}
  string.chars.to_a.uniq.each { |char|
    unique_chars_hash[char] = string.count(char)
    string.delete(char)
  }
  unique_chars_hash
end

string = 'Stringwithmultiplecharacters'
puts chars_counter_1(string)
puts chars_counter_2(string)

File: Python/009_string_each_char_occurrence.py
######################################################################
# this file contains different implementations of counting how many
# times each unique character appears in string
######################################################################


# simple solution with reducing size of original string
# during execution
def chars_counter_1(string):
    unique_chars_dict = {}
    uniques = ''.join(set(string))
    for char in uniques:
        unique_chars_dict[char] = string.count(char)
        string = string.replace(char, '')
    return unique_chars_dict


def chars_counter_2(string):
    return {char:string.count(char) for char in set(string)}


def main():
    string = 'Stringwithmultiplecharacters'
    print(chars_counter_1(string))
    print(chars_counter_2(string))

if __name__ == '__main__':
    main()